File: src/studio/lib/core/outcome.dart
/// 动作结果：命令行与窗口共用的一份算出来的东西。
///
/// `ok` 定退出码，`lines` 给命令行印，`columns` 与 `rows` 给窗口画；
/// 动作之间不互相打印，都只交出这一层。
class Outcome {
  Outcome(
    this.ok, {
    List<String>? lines,
    List<String>? columns,
    List<List<String>>? rows,
    this.payload,
  }) : lines = lines ?? <String>[],
       columns = columns ?? <String>[],
       rows = rows ?? <List<String>>[];

  Outcome.failed(List<String> lines) : this(false, lines: lines);

  bool ok;
  List<String> lines;
  List<String> columns;
  List<List<String>> rows;

  /// 有些动作要交出结构化的东西，而不是行列。
  Object? payload;

  /// 在最前面插一句。
  Outcome withFirst(String line) {
    lines.insert(0, line);
    return this;
  }

  Map<String, Object?> toJson() => payload != null && payload is Map
      ? (payload! as Map<String, Object?>)
      : {'ok': ok, 'lines': lines, 'columns': columns, 'rows': rows};

  @override
  String toString() => 'Outcome(ok: $ok, rows: ${rows.length})';
}

/// 路径相对根写短一点；不在根底下就原样。
String short(String root, String path) {
  final normalizedRoot = root.endsWith('/') ? root : '$root/';
  if (path.startsWith(normalizedRoot)) {
    return path.substring(normalizedRoot.length);
  }
  return path;
}

File: src/cli/src/outcome.rs
//! 结果：一次动作的答复。
//!
//! 规范（`docs/specification/process/outcome.md`）：一次动作一份，命令行与窗口都从它取，
//! 推进任务的动作还记进流水。这一份是它的不变部分——四样（`ok` 通不通、`lines` 话、
//! `columns` 与 `rows` 同一份表格、`data` 给界面的那一栏）与编解码；话怎么拼、
//! 路径怎么显示、退出码怎么定，留各自的平台。

use serde_json::{Value as Json, json};

/// 一次动作的答复。四样：通不通、话、表格、给界面的那一栏。
#[derive(Debug, Clone, Default, PartialEq)]
pub struct Outcome {
    pub ok: bool,
    /// 给人看的话，一行一句。
    pub lines: Vec<String>,
    /// 同一份表格：表头与行，命令行与窗口共用。
    pub columns: Vec<String>,
    pub rows: Vec<Vec<String>>,
    /// 给窗口与脚本的那一栏（要交原文就托在这里）；没有就不写。
    pub data: Option<Json>,
}

impl Outcome {
    pub fn new(ok: bool) -> Self {
        Outcome {
            ok,
            ..Default::default()
        }
    }

    /// 只有话的结果。
    pub fn lines(ok: bool, lines: Vec<String>) -> Self {
        Outcome {
            ok,
            lines,
            ..Default::default()
        }
    }

    /// 往话的开头添一句。
    pub fn with_first(mut self, line: String) -> Self {
        self.lines.insert(0, line);
        self
    }

    /// 信封：四样，`data` 有才写。
    pub fn to_json(&self) -> Json {
        let mut envelope = serde_json::Map::new();
        envelope.insert("ok".to_string(), json!(self.ok));
        envelope.insert("lines".to_string(), json!(self.lines));
        envelope.insert("columns".to_string(), json!(self.columns));
        envelope.insert("rows".to_string(), json!(self.rows));
        if let Some(data) = &self.data {
            envelope.insert("data".to_string(), data.clone());
        }
        Json::Object(envelope)
    }

    /// 原文那一栏（`--out` 落的就是它）；没托东西就给信封。
    pub fn to_output_json(&self) -> Json {
        match &self.data {
            Some(data) => data.clone(),
            None => self.to_json(),
        }
    }
}
